core/src/main/java/io/grpc/internal/TransportSet.java

```diff
@@ -233,15 +233,6 @@ private Runnable scheduleBackoff(
       final DelayedClientTransport delayedTransport, final Status status) {
     Preconditions.checkState(reconnectTask == null, "previous reconnectTask is not done");
```

Comment:
I think you need to move this check into scheduleEndOfCurrentBackoff(), which makes `scheduleBackoff()` do nothing other than returning a Runnable. This twisted monster is equivalent to its original form (before the changes made in #2154) called outside of synchronized in `transportShutdown()`.

Moving `scheduleBackoff()` outside of the synchronized block in `transportShutdown()` might be the better alternative to #2154. The lock in this case is to protect `reconnectPolicy` and `reconnectTask`, which are only changed in transport callbacks and when TransportSet is shutdown, thus nobody should change them between exiting the synchronized block and calling `scheduleBackoff()` in `transportShutdown()`.